helix-term/src/commands.rs

```diff
@@ -5100,3 +5102,240 @@ fn rename_symbol(cx: &mut Context) {
+fn decrement_number(cx: &mut Context) {
+    increment_number_impl(cx, -(cx.count() as i64));
+}
+
+/// If there is a number at the cursor, increment it by `amount`.
+fn increment_number_impl(cx: &mut Context, amount: i64) {
+    let (view, doc) = current!(cx.editor);
+
+    let selection = doc.selection(view.id);
+    if selection.len() != 1 {
+        return;
+    }
+
+    let primary_selection = selection.primary();
+    if primary_selection.to() - primary_selection.from() != 1 {
+        return;
+    }
+
+    let text = doc.text();
+    let cursor = primary_selection.cursor(text.slice(..));
+
+    if let Some(NumberInfo {
+        start,
+        end,
+        value: old_value,
+        radix,
+    }) = number_at(text, cursor)
+    {
+        let number_text: Cow<str> = text.slice(start..end).into();
+        let number_text = number_text.strip_prefix('-').unwrap_or(&number_text);
+
+        let new_value = old_value.wrapping_add(amount);
+        let old_length = end - start;
+
+        let (replacement, new_length) = {
+            let mut replacement = match radix {
+                2 => format!("{:b}", new_value),
+                8 => format!("{:o}", new_value),
+                10 => format!("{}", new_value.abs()),
+                16 => {
+                    let lower_count = number_text.chars().filter(char::is_ascii_lowercase).count();
+                    let upper_count = number_text.chars().filter(char::is_ascii_uppercase).count();
+                    if upper_count > lower_count {
+                        format!("{:X}", new_value)
+                    } else {
+                        format!("{:x}", new_value)
+                    }
+                }
+                _ => unimplemented!("radix not supported: {}", radix),
+            };
+
+            let mut new_length = replacement.chars().count();
+
+            let old_length_no_sign = number_text.chars().count();
+            if new_length < old_length_no_sign && (radix != 10 || number_text.starts_with('0')) {
+                replacement = "0".repeat(old_length_no_sign - new_length) + &replacement;
+                new_length = old_length_no_sign;
+            }
+
+            if radix == 10 && new_value.is_negative() {
+                replacement = format!("-{}", replacement);
+                new_length += 1;
+            }
+
+            (replacement, new_length)
+        };
+
+        let changes = std::iter::once((start, end, Some(replacement.into())));
+        let mut transaction = Transaction::change(text, changes);
+
+        // Move cursor to the last character of the number.
+        let mut selection = doc.selection(view.id).clone();
+        let mut primary = selection.primary_mut();
+        primary.anchor = if new_length < old_length {
+            end - 1 - old_length + new_length
+        } else {
+            end - 1 + new_length - old_length
+        };
+        primary.head = primary.anchor;
```

Comment:
I was trying to match the behavior of Vim, where incrementing/decrementing the number moves the cursor to the last character of the number, and I wanted this to be part of the transaction so that when you undo, the cursor moves back to where it was.

Also this implementation supports finding the next number on the line if the cursor isn't already over a number, which would move the cursor.

Maybe the way Vim does it doesn't make the most sense for helix. Would it make more sense to just have the number selected after the operation?